File: forms/advance-statement-about-care/full-stack-with-rust-axum-loco-tera-htmx-alpine/src/engine/completeness_grader.rs
use super::completeness_rules::all_rules;
use super::types::{AssessmentData, CompletenessStatus, FiredRule};
use super::utils::{count_completed_sections, determine_completeness};

/// Pure function: evaluates all completeness rules against assessment data.
/// Returns the completeness status, sections completed count, and fired rules.
pub fn calculate_completeness(data: &AssessmentData) -> (CompletenessStatus, u32, Vec<FiredRule>) {
    let sections_completed = count_completed_sections(data);
    let status = determine_completeness(data);

    // Fire rules
    let mut fired_rules = Vec::new();
    for rule in all_rules() {
        if (rule.evaluate)(data) {
            fired_rules.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                concern_level: rule.concern_level.to_string(),
            });
        }
    }

    (status.to_string(), sections_completed, fired_rules)
}

File: forms/advance-statement-about-care/front-end-form-with-svelte/src/lib/engine/completeness-grader.ts
import type { StatementData, CompletenessLevel, MissingSection, CompletenessResult } from './types';
import { completenessRules } from './completeness-rules';
import { detectFlaggedIssues } from './flagged-issues';

/**
 * Pure function: evaluates all completeness rules against statement data.
 * Returns a completeness level based on how many sections are filled:
 *   - Incomplete: critical required sections are missing
 *   - Partial: some sections completed but required ones missing
 *   - Complete: all required sections filled
 *   - Verified: complete + witnessed + signed by healthcare professional
 */
export function calculateCompleteness(data: StatementData): CompletenessResult {
	const missingSections: MissingSection[] = [];
	let completedCount = 0;
	const totalCount = completenessRules.length;

	for (const rule of completenessRules) {
		try {
			if (rule.evaluate(data)) {
				completedCount++;
			} else {
				missingSections.push({
					id: rule.id,
					section: rule.section,
					description: rule.description,
					required: rule.required
				});
			}
		} catch (e) {
			// Rule evaluation failed - treat as missing
			console.warn(`Completeness rule ${rule.id} evaluation failed:`, e);
			missingSections.push({
				id: rule.id,
				section: rule.section,
				description: rule.description,
				required: rule.required
			});
		}
	}

	const flaggedIssues = detectFlaggedIssues(data);
	const missingRequired = missingSections.filter((s) => s.required);
	const level = determineLevel(data, missingRequired.length, completedCount);

	return {
		level,
		missingSections,
		flaggedIssues,
		completedCount,
		totalCount,
		timestamp: new Date().toISOString()
	};
}

function determineLevel(
	data: StatementData,
	missingRequiredCount: number,
	completedCount: number
): CompletenessLevel {
	// Verified: all required complete + witnessed + healthcare professional acknowledged
	const isWitnessed =
		data.signaturesWitnesses.witnessName.trim() !== '' &&
		data.signaturesWitnesses.witnessSignature.trim() !== '';
	const isProfessionalAcknowledged =
		data.signaturesWitnesses.healthcareProfessionalName.trim() !== '' &&
		data.signaturesWitnesses.healthcareProfessionalSignature.trim() !== '';

	if (missingRequiredCount === 0 && isWitnessed && isProfessionalAcknowledged) {
		return 'verified';
	}

	// Complete: all required sections filled
	if (missingRequiredCount === 0) {
		return 'complete';
	}

	// Partial: at least some sections completed
	if (completedCount > 0) {
		return 'partial';
	}

	// Incomplete: nothing filled
	return 'incomplete';
}
